bqskit/passes/search/generators/discrete.py

```diff
@@ -0,0 +1,206 @@
+"""This module implements the DiscreteLayerGenerator class."""
+from __future__ import annotations
+
+import logging
+from typing import Sequence
+
+from bqskit.compiler.passdata import PassData
+from bqskit.ir.circuit import Circuit
+from bqskit.ir.gate import Gate
+from bqskit.ir.gates.constant.cx import CNOTGate
+from bqskit.ir.gates.constant.h import HGate
+from bqskit.ir.gates.constant.t import TGate
+from bqskit.ir.gates.parameterized.pauliz import PauliZGate
+from bqskit.ir.operation import Operation
+from bqskit.passes.search.generator import LayerGenerator
+from bqskit.qis.state.state import StateVector
+from bqskit.qis.state.system import StateSystem
+from bqskit.qis.unitary.unitarymatrix import UnitaryMatrix
+from bqskit.utils.typing import is_sequence
+
+
+_logger = logging.getLogger(__name__)
+
+
+class DiscreteLayerGenerator(LayerGenerator):
+    """
+    The DiscreteLayerGenerator class.
+
+    Expands circuits using only discrete gates.
+    """
+
+    def __init__(
+        self,
+        gateset: Sequence[Gate] = [HGate(), TGate(), CNOTGate()],
+        double_headed: bool = False,
+    ) -> None:
+        """
+        Construct a DiscreteLayerGenerator.
+
+        Args:
+            gateset (Sequence[Gate]): A sequence of gates that can be used
+                in the output circuit. These must be non-parameterized gates.
+                (Default: [HGate, TGate, CNOTGate])
+
+            double_headed (bool): If True, successors will be generated by
+                both appending and prepending gates. (Default: False)
+
+        Raises:
+            TypeError: If the gateset is not a sequence.
+
+            TypeError: If the gateset contains a parameterized gate.
+
+            TypeError: If the radices of gates are different.
```

Comment:
This is worth discussing more.

The policy as it is now is to only allow single radix circuits/targets. The reason being is that if there are two different radices, then it might be the case that two universal gate sets (one for each radix) need to be included in `gateset`. Otherwise someone might try to synthesize a multi-radix target and end up waiting forever.

Checking if a get set does not seem to be very easy to do (https://www.scottaaronson.com/qclec/16.pdf), but verifying that a gate set is probably not universal can be done.

The question is then: is it better to allow for multi-radix targets here and assume the user is supplying universal gate sets for both?

Reply:
It isn't the layer generator's job to determine universality. There are legitimate cases where non-universal gate sets might be used with a layer generator. So, I would go with your last point: assume the user knows what they are doing. However, I think our synthesis algorithms should always print warnings for no progress to help inform the user of a mistake like this.

Additionally, it would be good to have the code to check for (not) universality somewhere inside BQSKit, maybe inside the gateset object. I do believe we could concoct a simple universality test via synthesis, although it would be slower than analytical solutions.

Reply:
I'm going to add this as a separate issue as a feature request because it's pretty non-trivial and useful beyond this context.